### PR TITLE
Introduce additional widgets

### PR DESCRIPTION
**Pull request by pasnox**
- QShape
- QUpDownButton

### CODE REVIEW

File: qtwidgetsextra/additional/QShape.cpp [hunk 1]
```diff
@@ -0,0 +1,253 @@
+#include "QShape.h"
+
+class QShapePrivate : public QObject {
+    Q_OBJECT
+
+public:
+    QShapePrivate(QShape *widgetP)
+        : QObject(widgetP)
+        , widget(widgetP)
+        , renderHints(QPainter::TextAntialiasing /*| QPainter::Antialiasing*/)
```
Comment (Schoumi) [inline qtwidgetsextra/additional/QShape.cpp]: delete unnecessary comment

File: qtwidgetsextra/additional/QShape.cpp [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+QShape::QShape(QShape::Shape shape, QWidget *parent)
+    : QFrame(parent)
+    , d(new QShapePrivate(this))
+{
+    setShape(shape);
+}
+
+QSize QShape::minimumSizeHint() const
+{
+    return d->shape == QShape::User ? d->userPath.boundingRect().size().toSize() :QSize(30,30);
```
Comment (Schoumi) [inline qtwidgetsextra/additional/QShape.cpp]: Ternary if made readability much complexe than necessary

File: qtwidgetsextra/additional/QShape.cpp [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+}
+
+QSize QShape::sizeHint() const
+{
+    return d->shape == QShape::User ? d->userPath.boundingRect().size().toSize() :QSize(50,50);
```
Comment (Schoumi) [inline qtwidgetsextra/additional/QShape.cpp]: Ternary if made readability much complexe than necessary

File: qtwidgetsextra/additional/QShape.h [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    Q_PROPERTY(QPainter::RenderHints renderHints READ renderHints WRITE setRenderHints)
+    Q_PROPERTY(QMargins margins READ margins WRITE setMargins)
+    Q_PROPERTY(QBrush brush READ brush WRITE setBrush)
+    Q_PROPERTY(QPen pen READ pen WRITE setPen)
+    Q_PROPERTY(QShape::Shape shape READ shape WRITE setShape)
+    Q_PROPERTY(QPoint radius READ radius WRITE setRadius)
+
+    enum ShapeFlag {
+        CircleFlag = 0x1,
+        DiamondFlag = 0x2,
```
Comment (Schoumi) [inline qtwidgetsextra/additional/QShape.h]: why hex values and it's better to have it in the right order. especially if this has to grow. Two enum values can have the same associated values

File: qtwidgetsextra/additional/QShapePlugin.cpp [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    , m_initialized(false)
+{
+}
+
+void QShapePlugin::initialize(QDesignerFormEditorInterface * /* core */)
+{
+    if (m_initialized) {
+        return;
+    }
+
```
Comment (Schoumi) [inline qtwidgetsextra/additional/QShapePlugin.cpp]: too empty lines

File: qtwidgetsextra/additional/QUpDownButtonPlugin.cpp [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    : QObject(parent)
+    , m_initialized(false)
+{
+}
+
+void QUpDownButtonPlugin::initialize(QDesignerFormEditorInterface * /* core */)
+{
+    if (m_initialized) {
+        return;
+    }
```
Comment (Schoumi) [inline qtwidgetsextra/additional/QUpDownButtonPlugin.cpp]: too empty lines

File: qtwidgetsextra/additional/QShape.h (hunk 4 above)
Comment (pasnox) [inline qtwidgetsextra/additional/QShape.h]: That's an or'derable flag, not a enum.

File: qtwidgetsextra/additional/QShape.cpp (hunk 1 above)
Comment (pasnox) [reply qtwidgetsextra/additional/QShape.cpp]: done

File: qtwidgetsextra/additional/QShape.cpp (hunk 2 above)
Comment (pasnox) [reply qtwidgetsextra/additional/QShape.cpp]: done

File: qtwidgetsextra/additional/QShape.cpp (hunk 3 above)
Comment (pasnox) [reply qtwidgetsextra/additional/QShape.cpp]: done

File: qtwidgetsextra/additional/QShapePlugin.cpp (hunk 5 above)
Comment (pasnox) [reply qtwidgetsextra/additional/QShapePlugin.cpp]: done

File: qtwidgetsextra/additional/QUpDownButtonPlugin.cpp (hunk 6 above)
Comment (pasnox) [reply qtwidgetsextra/additional/QUpDownButtonPlugin.cpp]: done

**Pull request closed without merging**